diff --git a/src/vte.cc b/src/vte.cc
index 5f2c8f82..24247907 100644
--- a/src/vte.cc
+++ b/src/vte.cc
@@ -10138,6 +10138,12 @@ Terminal::widget_mouse_scroll(vte::platform::ScrollEvent const& event)
 	gdouble v;
         gint cnt_x, cnt_y, i;
 	int button;
+        auto const quantise_surface_scroll = [](double* delta,
+                                                double step) noexcept {
+                auto const count = int(*delta / step);
+                *delta -= count * step;
+                return count;
+        };
 
         m_modifiers = event.modifiers();
         m_mouse_smooth_scroll_x_delta += event.dx();
@@ -10145,19 +10151,27 @@ Terminal::widget_mouse_scroll(vte::platform::ScrollEvent const& event)
 
         /* If we're running a mouse-aware application, map the scroll event to button presses on buttons 4-7. */
 	if (m_mouse_tracking_mode != MouseTrackingMode::eNONE) {
-                cnt_x = m_mouse_smooth_scroll_x_delta;
-                cnt_y = m_mouse_smooth_scroll_y_delta;
+                if (event.is_precise()) {
+                        cnt_x = quantise_surface_scroll(&m_mouse_smooth_scroll_x_delta,
+                                                        MAX(1., double(m_cell_width)));
+                        cnt_y = quantise_surface_scroll(&m_mouse_smooth_scroll_y_delta,
+                                                        MAX(1., double(m_cell_height)));
+                } else {
+                        cnt_x = m_mouse_smooth_scroll_x_delta;
+                        cnt_y = m_mouse_smooth_scroll_y_delta;
+                        m_mouse_smooth_scroll_x_delta -= cnt_x;
+                        m_mouse_smooth_scroll_y_delta -= cnt_y;
+                }
                 if (cnt_x == 0 && cnt_y == 0)
 			return true;
 
                 /* Need to ensure the ringview is updated. */
                 ringview_update();
 
-                m_mouse_smooth_scroll_x_delta -= cnt_x;
-                m_mouse_smooth_scroll_y_delta -= cnt_y;
 		_vte_debug_print(VTE_DEBUG_EVENTS,
-                                 "Scroll application by %d lines, %d columns, smooth scroll delta set back to y=%f, x=%f\n",
-                                 cnt_y, cnt_x, m_mouse_smooth_scroll_y_delta, m_mouse_smooth_scroll_x_delta);
+                                 "Scroll application by %d lines, %d columns, precise=%d, smooth scroll delta set back to y=%f, x=%f\n",
+                                 cnt_y, cnt_x, event.is_precise(),
+                                 m_mouse_smooth_scroll_y_delta, m_mouse_smooth_scroll_x_delta);
 
                 button = cnt_y > 0 ? 5 : 4;
                 if (cnt_y < 0)
@@ -10195,13 +10209,19 @@ Terminal::widget_mouse_scroll(vte::platform::ScrollEvent const& event)
 		char *normal;
 		gsize normal_length;
 
-                cnt_y = v * m_mouse_smooth_scroll_y_delta;
+                if (event.is_precise()) {
+                        cnt_y = quantise_surface_scroll(&m_mouse_smooth_scroll_y_delta,
+                                                        MAX(1., double(m_cell_height)));
+                } else {
+                        cnt_y = v * m_mouse_smooth_scroll_y_delta;
+                        if (cnt_y != 0)
+                                m_mouse_smooth_scroll_y_delta -= cnt_y / v;
+                }
                 if (cnt_y == 0)
 			return true;
-                m_mouse_smooth_scroll_y_delta -= cnt_y / v;
 		_vte_debug_print(VTE_DEBUG_EVENTS,
-                                "Scroll by %d lines, smooth scroll y_delta set back to %f\n",
-                                cnt_y, m_mouse_smooth_scroll_y_delta);
+                                "Scroll by %d lines, precise=%d, smooth scroll y_delta set back to %f\n",
+                                cnt_y, event.is_precise(), m_mouse_smooth_scroll_y_delta);
 
                 /* In the alternate screen there is no scrolling, so fake a few cursor keystrokes (only vertically). */
 		_vte_keymap_map (
@@ -10228,6 +10248,13 @@ Terminal::widget_mouse_scroll(vte::platform::ScrollEvent const& event)
         return false;
 }
 
+void
+Terminal::reset_mouse_smooth_scroll() noexcept
+{
+        m_mouse_smooth_scroll_x_delta = 0.;
+        m_mouse_smooth_scroll_y_delta = 0.;
+}
+
 bool
 Terminal::set_audible_bell(bool setting)
 {
diff --git a/src/vteinternal.hh b/src/vteinternal.hh
index 4d26a4f9..5ca77ee4 100644
--- a/src/vteinternal.hh
+++ b/src/vteinternal.hh
@@ -1108,6 +1108,7 @@ public:
         void widget_mouse_enter(vte::platform::MouseEvent const& event);
         void widget_mouse_leave(vte::platform::MouseEvent const& event);
         bool widget_mouse_scroll(vte::platform::ScrollEvent const& event);
+        void reset_mouse_smooth_scroll() noexcept;
 #if VTE_GTK == 4
         bool widget_key_modifiers(unsigned modifiers);
 #endif /* VTE_GTK == 4 */
diff --git a/src/widget.cc b/src/widget.cc
index 0daaed1e..d4559af7 100644
--- a/src/widget.cc
+++ b/src/widget.cc
@@ -1245,7 +1245,8 @@ Widget::event_scroll_begin(GtkEventControllerScroll* controller)
 {
         _vte_debug_print(VTE_DEBUG_EVENTS, "Scroll begin\n");
 
-        // FIXMEgtk4
+        if (gtk_event_controller_scroll_get_unit(controller) == GDK_SCROLL_UNIT_SURFACE)
+                terminal()->reset_mouse_smooth_scroll();
 }
 
 bool
@@ -1260,6 +1261,9 @@ Widget::event_scroll(GtkEventControllerScroll* controller,
                 return false;
 
         return terminal()->widget_mouse_scroll({gdk_event_get_modifier_state(event),
+                                                gtk_event_controller_scroll_get_unit(controller) == GDK_SCROLL_UNIT_SURFACE
+                                                        ? ScrollEvent::Unit::eSURFACE
+                                                        : ScrollEvent::Unit::eWHEEL,
                                                 dx, dy});
 }
 
@@ -1268,7 +1272,8 @@ Widget::event_scroll_end(GtkEventControllerScroll* controller)
 {
         _vte_debug_print(VTE_DEBUG_EVENTS, "Scroll end\n");
 
-        // FIXMEgtk4
+        if (gtk_event_controller_scroll_get_unit(controller) == GDK_SCROLL_UNIT_SURFACE)
+                terminal()->reset_mouse_smooth_scroll();
 }
 
 void
@@ -1758,6 +1763,7 @@ Widget::scroll_event_from_gdk(GdkEvent* event) const
         if (gdk_event_get_pointer_emulated(event))
                 return std::nullopt;
 
+        auto unit = ScrollEvent::Unit::eWHEEL;
         auto dx = double{}, dy = double{};
         if (!gdk_event_get_scroll_deltas(event, &dx, &dy)) {
                 auto dir = GdkScrollDirection{};
@@ -1772,10 +1778,12 @@ Widget::scroll_event_from_gdk(GdkEvent* event) const
                 case GDK_SCROLL_SMOOTH:
                 default: __builtin_unreachable();
                 }
+        } else {
+                unit = ScrollEvent::Unit::eSURFACE;
         }
 
 
-        return ScrollEvent{read_modifiers_from_gdk(event), dx, dy};
+        return ScrollEvent{read_modifiers_from_gdk(event), unit, dx, dy};
 }
 
 #endif /* VTE_GTK == 3 */
diff --git a/src/widget.hh b/src/widget.hh
index 4dd78c2c..93cc7ba5 100644
--- a/src/widget.hh
+++ b/src/widget.hh
@@ -226,13 +226,20 @@ class ScrollEvent : public EventBase {
 
 protected:
 
+        enum class Unit {
+                eWHEEL,
+                eSURFACE,
+        };
+
         ScrollEvent() noexcept = default;
 
         constexpr ScrollEvent(unsigned modifiers,
+                              Unit unit,
                               double dx,
                               double dy) noexcept
                 : EventBase{EventBase::Type::eMOUSE_SCROLL},
                   m_modifiers{modifiers},
+                  m_unit{unit},
                   m_dx{dx},
                   m_dy{dy}
         {
@@ -247,11 +254,14 @@ public:
         ScrollEvent& operator=(ScrollEvent&&) = delete;
 
         constexpr auto modifiers() const noexcept { return m_modifiers; }
+        constexpr auto unit() const noexcept { return m_unit; }
+        constexpr auto is_precise() const noexcept { return m_unit == Unit::eSURFACE; }
         constexpr auto dx()        const noexcept { return m_dx;        }
         constexpr auto dy()        const noexcept { return m_dy;        }
 
 private:
         unsigned m_modifiers;
+        Unit m_unit{Unit::eWHEEL};
         double m_dx;
         double m_dy;
 }; // class ScrollEvent
